geofence.py

```diff
@@ -35,52 +38,100 @@ def Y(self) -> float:
+        return f"(lat={self.latitude};long={self.longitude})"
+
+
+class TileWrapper:
+    def __init__(self, tile: Tile):
+        self.tile = tile
+        self.fullname = f"'{self.tile.name}' ({self.tile.uuid})"
+        self.previouslyIgnored: bool = False
+        self._location: Point = None
+
+    async def refresh(self, api) -> None:
+        self.tile._async_request = api._async_request
+        await self.tile.async_update()
+        self._location = None
+
+    @property
+    def location(self) -> Point:
+        if not self._location:
+            self._location = Point(longitude=self.tile.longitude, latitude=self.tile.latitude)
+        return self._location
+
+    @property
+    def last_timestamp(self) -> Optional[datetime]:
+        return self.tile.last_timestamp
+
+    @property
+    def name(self) -> str:
+        return self.tile.name
+
+    @property
+    def uuid(self) -> str:
+        return self.tile.uuid
+
+
+class GeoConfig:
+    def __init__(self, hubitat_devices: dict[int, str]) -> None:
+        self.all_tiles: set[str] = set()
+        self.hubitatIds: set[int] = set()
+        self.hubitat_devices = hubitat_devices
 
 
 class Geofence:
 
-    def __init__(self, name: str, conf: dict, all_tiles: set[str], hubitatIds: set[int], hubitat_devices: dict[int, str]) -> None:
+    def __init__(self, name: str, conf: dict, geoconf: GeoConfig, exclusion: bool) -> None:
         self.name = name
+        self.exclusion = exclusion
+        key = "tiles"
 
-        if "tile_to_hubitat" not in conf:
-            logging.warn(f"No tiles or Hubitat devices associated with section '{name}'.")
+        if key not in conf:
+            logging.warn(f"No tiles associated with location '{name}'.")
             return
 
-        self.tiles = conf["tile_to_hubitat"]
+        self.tiles = {x: None for x in conf[key]} if exclusion else conf[key]
 
         for tile, hubitatId in self.tiles.items():
-            all_tiles.add(tile)
+            geoconf.all_tiles.add(tile)
+            if not hubitatId:
+                continue
             hubitatId = int(hubitatId)
-            if hubitatId in hubitatIds:
-                raise Exception(f"Hubitat device Id {hubitatId} is referenced in section '{name}' and another section.")
-            hubitatIds.add(hubitatId)
-            if hubitatId not in hubitat_devices:
+            if hubitatId in geoconf.hubitatIds:
+                raise Exception(f"Hubitat device Id {hubitatId} is referenced in location '{name}' and another location.")
+            geoconf.hubitatIds.add(hubitatId)
+            if not hubitatId in geoconf.hubitat_devices:
                 raise Exception(f"Hubitat device Id {hubitatId} is not a virtual presence sensor exported by Hubitat's MakerAPI.")
 
     def isInside(self, p: Point) -> bool:
         return False
 
-    def processTile(self, p: Point, name: str, uuid: str, hubitat: Hubitat) -> None:
+    def processTile(self, tile: TileWrapper, hubitat: Hubitat) -> bool:
```

Comment:
Having `tile` as `TileWrapper` is weird.  Consider:
- `tile_wrapper`
- rename  `Tile` to `TileBase`, `TileWrapper` to `Tile`

Reply:
Unfortunately, Tile is coming from the pytile package.

Reply:
ok!